src/Mvc/Controller/AbstractRestfulController.php

```diff
@@ -0,0 +1,163 @@
+        }
+
+        $routeParameters = $this->params()->fromRoute(null, []);
+        unset($routeParameters['controller'], $routeParameters['action']);
+
+        $result = $this->$method($routeParameters);
+
+        $event->setResult($result);
+    }
+
+    /**
+     * Provides a built-in implementation for the HTTP OPTIONS method
+     *
+     * It returns the supported methods for the given resource
+     *
+     * @return HttpResponse
+     */
+    public function options()
+    {
+        /** @var HttpResponse $response */
+        $response = $this->getResponse();
+
+        $response->setContent('');
+        $response->setStatusCode(200);
+        $response->getHeaders()->addHeaderLine('Allow', implode(', ', $this->getAllowedVerbs()));
+
+        return $response;
+    }
+
+    /**
+     * @TODO: when we have the new input filter, we should use named validation group and context
+     *
+     * @param  string $inputFilterName
+     * @param  array  $validationGroup
+     * @return array
+     * @throws
+     */
+    protected function validateIncomingData($inputFilterName, array $validationGroup = [])
+    {
+        /** @var \Zend\InputFilter\InputFilterInterface $inputFilter */
+        $inputFilter = $this->serviceLocator->get('InputFilterManager')->get($inputFilterName);
+
+        if (!empty($validationGroup)) {
+            $inputFilter->setValidationGroup($validationGroup);
+        }
+
+        $data = json_decode($this->request->getContent(), true);
+        $inputFilter->setData($data);
+
+        if ($inputFilter->isValid()) {
+            return $inputFilter->getValues();
+        }
+
+        throw new UnprocessableEntityException('Validation error', $this->extractErrorMessages($inputFilter));
+    }
+
+    /**
+     * @param  string $hydratorName
+     * @param  object $object
+     * @param  array  $values
+     * @return object
+     */
+    protected function hydrateData($hydratorName, $object, array $values)
+    {
+        /** @var \Zend\Stdlib\Hydrator\HydratorInterface $hydrator */
+        $hydrator = $this->serviceLocator->get('HydratorManager')->get($hydratorName);
+        $object   = $hydrator->hydrate($values, $object);
+
+        return $object;
+    }
+
+    /**
+     * @param  InputFilterInterface $inputFilter
+     * @return array
+     */
+    private function extractErrorMessages(InputFilterInterface $inputFilter)
+    {
+        $errorMessages = $inputFilter->getMessages();
+
+        array_walk($errorMessages, function (&$value, $key) use ($inputFilter) {
+            if ($inputFilter->has($key) && $inputFilter->get($key) instanceof InputFilterInterface) {
+                $value = $this->extractErrorMessages($inputFilter->get($key));
+            } else {
+                $value = array_values($value);
+            }
+        });
+
+        return $errorMessages;
+    }
+
+    /**
+     * @return array
+     */
+    private function getAllowedVerbs()
```

Comment:
Can you make this method protected? So that I could override this method and hardcode my allowed methods